File: src/main/java/ru/octol1ttle/flightassistant/computers/impl/safety/StallComputer.java
package ru.octol1ttle.flightassistant.computers.impl.safety;

import net.minecraft.util.math.Direction;
import net.minecraft.util.math.MathHelper;
import ru.octol1ttle.flightassistant.computers.api.ControlInput;
import ru.octol1ttle.flightassistant.computers.api.IPitchLimiter;
import ru.octol1ttle.flightassistant.computers.api.IThrustController;
import ru.octol1ttle.flightassistant.computers.api.ITickableComputer;
import ru.octol1ttle.flightassistant.computers.api.InputPriority;
import ru.octol1ttle.flightassistant.computers.impl.AirDataComputer;
import ru.octol1ttle.flightassistant.config.ComputerConfig;
import ru.octol1ttle.flightassistant.config.FAConfig;
import ru.octol1ttle.flightassistant.registries.ComputerRegistry;

public class StallComputer implements ITickableComputer, IPitchLimiter, IThrustController {
    private final AirDataComputer data = ComputerRegistry.resolve(AirDataComputer.class);
    public StallStatus status = StallStatus.UNKNOWN;
    public float maximumSafePitch = 90.0f;

    @Override
    public void tick() {
        status = computeStalling();
        maximumSafePitch = computeMaximumSafePitch();
    }

    private StallStatus computeStalling() {
        if (!data.isFlying() || data.player().isTouchingWater()) {
            return StallStatus.UNKNOWN;
        }
        if (data.isInvulnerableTo(data.player().getDamageSources().fall())) {
            return StallStatus.PLAYER_INVULNERABLE;
        }
        if (data.fallDistance() <= 3.0f) {
            return StallStatus.FALL_DISTANCE_TOO_LOW;
        }
        if (data.pitch() - data.flightPitch <= 45.0f) {
            return StallStatus.AOA_SAFE;
        }
        if (data.pitch() - data.flightPitch <= 90.0f || data.velocity.y > -10.0f) {
            return StallStatus.APPROACHING_STALL;
        }
        return StallStatus.FULL_STALL;
    }

    private float computeMaximumSafePitch() {
        if (!data.isFlying() || status == StallStatus.UNKNOWN || status == StallStatus.PLAYER_INVULNERABLE) {
            return 90.0f;
        }
        return MathHelper.clamp(data.flightPitch + 90.0f, 0.0f, 90.0f);
    }

    @Override
    public float getMaximumPitch() {
        return maximumSafePitch;
    }

    @Override
    public boolean blockPitchChange(Direction direction) {
        return direction == Direction.UP && status == StallStatus.FULL_STALL;
    }

    @Override
    public ComputerConfig.ProtectionMode getProtectionMode() {
        return FAConfig.computer().stallProtection;
    }

    @Override
    public ControlInput getThrustInput() {
        if (!FAConfig.computer().stallUseThrust || status != StallStatus.FULL_STALL) {
            return null;
        }
        return new ControlInput(1.0f, 1.0f, InputPriority.HIGHEST);
    }

    @Override
    public String getFaultTextBaseKey() {
        return "alerts.flightassistant.fault.computers.stall_det";
    }

    @Override
    public void reset() {
        status = StallStatus.UNKNOWN;
        maximumSafePitch = 90.0f;
    }

    public enum StallStatus {
        FULL_STALL,
        APPROACHING_STALL,
        AOA_SAFE,
        FALL_DISTANCE_TOO_LOW,
        PLAYER_INVULNERABLE,
        UNKNOWN
    }
}

File: src/main/kotlin/ru/octol1ttle/flightassistant/impl/computer/safety/StallComputer.kt
package ru.octol1ttle.flightassistant.impl.computer.safety

import net.minecraft.text.Text
import net.minecraft.util.Identifier
import ru.octol1ttle.flightassistant.FlightAssistant
import ru.octol1ttle.flightassistant.api.computer.Computer
import ru.octol1ttle.flightassistant.api.computer.ComputerAccess
import ru.octol1ttle.flightassistant.api.autoflight.ControlInput
import ru.octol1ttle.flightassistant.api.autoflight.pitch.PitchLimiter
import ru.octol1ttle.flightassistant.api.autoflight.thrust.ThrustController
import ru.octol1ttle.flightassistant.api.autoflight.pitch.PitchLimiterRegistrationCallback
import ru.octol1ttle.flightassistant.api.autoflight.thrust.ThrustControllerRegistrationCallback
import ru.octol1ttle.flightassistant.api.util.extensions.data
import ru.octol1ttle.flightassistant.config.FAConfig

class StallComputer : Computer(), PitchLimiter, ThrustController {
    var status: Status = Status.SAFE
        private set
    private var maximumSafePitch: Float = 90.0f

    override fun subscribeToEvents() {
        PitchLimiterRegistrationCallback.EVENT.register { it.accept(this) }
        ThrustControllerRegistrationCallback.EVENT.register { it.accept(this) }
    }

    override fun tick(computers: ComputerAccess) {
        status =
            if (computers.data.flying && computers.data.velocity.y < 0 && !computers.data.fallDistanceSafe && computers.data.forwardVelocity.length() == 0.0)
                if (status == Status.FULL_STALL || computers.data.velocity.y * 20 <= -10) Status.FULL_STALL else Status.APPROACHING_STALL
            else Status.SAFE
        maximumSafePitch = (computers.data.forwardVelocity.length() * 20.0f * 3.0f).toFloat().coerceIn(0.0f..90.0f)
    }

    override fun getThrustInput(computers: ComputerAccess): ControlInput? {
        if (FAConfig.safety.stallAutoThrust && status != Status.SAFE) {
            return ControlInput(
                1.0f,
                ControlInput.Priority.HIGHEST,
                Text.translatable("mode.flightassistant.thrust.toga"),
                active = status == Status.FULL_STALL
            )
        }

        return null
    }

    override fun getMaximumPitch(computers: ComputerAccess): ControlInput? {
        if (computers.data.fallDistanceSafe) {
            return null
        }

        return ControlInput(
            maximumSafePitch,
            ControlInput.Priority.HIGHEST,
            Text.translatable("mode.flightassistant.pitch.stall_protection"),
            active = FAConfig.safety.stallLimitPitch
        )
    }

    override fun reset() {
        status = Status.SAFE
        maximumSafePitch = 90.0f
    }

    enum class Status {
        FULL_STALL,
        APPROACHING_STALL,
        SAFE
    }

    companion object {
        val ID: Identifier = FlightAssistant.id("stall")
    }
}
